xls/dslx/fmt/ast_fmt.cc

```diff
@@ -1234,9 +1237,34 @@ static DocRef FmtStructMembersBreak(
           return arena.MakeText(name);
         }
 
-        return ConcatNGroup(arena,
-                            {arena.MakeText(name), arena.colon(), arena.space(),
-                             arena.MakeGroup(Fmt(*expr, comments, arena))});
+        DocRef field_expr = Fmt(*expr, comments, arena);
+
+        // This is the document we want to emit both when we:
+        // - Know it fits in flat mode
+        // - Know the start of the document (i.e. the leader on the RHS
+        //   expression) can be emitted in flat mode
+        //
+        // That's why it has a `break1` in it (instead of a space) and a
+        // reassessment of whether to enter break mode for the field
+        // expression.
+        DocRef on_flat =
+            ConcatN(arena, {arena.MakeText(name), arena.colon(), arena.break1(),
+                            arena.MakeGroup(field_expr)});
+        DocRef nest_field_expr =
+            ConcatN(arena, {arena.MakeText(name), arena.colon(),
+                            arena.hard_line(), arena.MakeNest(field_expr)});
+
+        DocRef on_other;
+        if (expr->IsBlockedExprWithLeader()) {
+          DocRef leader = arena.MakeConcat(
+              arena.space(), FmtBlockedExprLeader(*expr, comments, arena));
```

Comment:
Yeah I need to add more formatting tests to show that the "leader" is an exact match to the amount of space left in the line -- technically being conservative by one space doesn't hurt anything, but we should be accurate.

Note that the "leader" doc is just used to select the mode, if it fits then "flat" mode is used to emit the subsequent doc, otherwise "break" mode is used. See the PR description of https://github.com/google/xls/pull/1620 for a pretty lengthy explanation -- will add tests to ensure accuracy in the meantime.

Reply:
Ok added tests for exactly 100 chars and one-over 100 chars.

Reply:
(Before the field name wasn't being included so it was worse than just being conservative, it was inaccurate :+1:  )